fix: variable name changes when element name/label changes

abdul99ahad opened the pull request:
Closes #863

### Proposed Changes

- [x] When creating or updating the label in Decision box, the variable name automatically updates 
- [x] Setting the element name automatically sets variable name 

<!--

Thanks for creating this pull request! ❤️

-->

barmac commented:
I don't see any changes regarding the decision table. Is it already present in that editor? I know that we don't support variable adding or modification in the decision table, but it could be present in the XML brought by user.

abdul99ahad commented:
> I don't see any changes regarding the decision table. Is it already present in that editor? I know that we don't support variable adding or modification in the decision table, but it could be present in the XML brought by user.

Since variables weren't available in the modeling interface, I initially assumed they weren't part of the decision tables. However, I've now fixed the issue for all elements to ensure everything is covered.

barmac commented on packages/dmn-js-shared/src/features/modeling/behavior/NameChangeBehavior.js:
```diff
@@ -0,0 +1,87 @@
+      return;
+    }
+
+    bo.variable.name = newLabel;
+    this._modeling.updateProperties(element, { variable: bo.variable });
```
This is a technique called monkey-patching. If you add tests for `commandStack.undo`, you will notice that the change cannot be undone. We need to dispatch an updateProperties call which updates the variable name.
@philippfromme perhaps it makes sense you look into this together with @abdul99ahad

abdul99ahad replied:
Thanks, @barmac, for pointing this out. I discussed the monkey patching technique with @philippfromme earlier and I didn't realize that my implementation went against the recommended approach. I apologize for the oversight. I've updated the implementation to align with the correct approach.

barmac reviewed (changes requested):
Notice the comment I left. Let's add tests for undo, and fix the code. As I am off for the next two weeks, please ask @philippfromme or other team member for support.

abdul99ahad commented:
> Notice the comment I left. Let's add tests for undo, and fix the code. As I am off for the next two weeks, please ask @philippfromme or other team member for support.

I have fixed the code and added the test cases that support undo/redo

abdul99ahad commented:
@barmac @philippfromme We could move the `NameChangeBehavior` from dmn-boxed-expression to a shared `NameChangeBehavior`. The key difference is that the shared version only syncs the element name with the variable name, not the other way around, while the boxed-expression version supports both directions. Should we keep this distinction, or do we need to refactor?

philippfromme commented:
> @barmac @philippfromme We could move the `NameChangeBehavior` from dmn-boxed-expression to a shared `NameChangeBehavior`. The key difference is that the shared version only syncs the element name with the variable name, not the other way around, while the boxed-expression version supports both directions. Should we keep this distinction, or do we need to refactor?

Is there anything keeping us from handling both ways?

abdul99ahad commented:
> > @barmac @philippfromme We could move the `NameChangeBehavior` from dmn-boxed-expression to a shared `NameChangeBehavior`. The key difference is that the shared version only syncs the element name with the variable name, not the other way around, while the boxed-expression version supports both directions. Should we keep this distinction, or do we need to refactor?
> 
> Is there anything keeping us from handling both cases in a shared component?

We can cover that case as well. I previously discussed this with @barmac, who mentioned that only a one-way sync is required as of now.

nikku commented on packages/dmn-js-decision-table/test/spec/features/modeling/behavior/NameChangeBehaviorSpec.js:
```diff
@@ -0,0 +1,86 @@
+import Modeling from 'src/features/modeling';
+
+
+describe('NameChangeBehavior', function() {
```
```suggestion
describe('features/modeling/behavior - NameChangeBehavior', function() {
```

barmac commented:
I will look into this today again.

barmac commented:
Note that the variable name change is not reflected in the literal expression editor:

https://github.com/user-attachments/assets/52f9448e-2267-409b-b609-9740f2b20bff

barmac commented:
Looks that it's an already existing issue. https://github.com/bpmn-io/dmn-js/blob/c3e093984603e14e91daecb5cc11c9f34b3e7ff9/packages/dmn-js-literal-expression/src/features/literal-expression-properties/components/LiteralExpressionPropertiesEditorComponent.js does not update on external changes.

barmac commented:
> > > @barmac @philippfromme We could move the `NameChangeBehavior` from dmn-boxed-expression to a shared `NameChangeBehavior`. The key difference is that the shared version only syncs the element name with the variable name, not the other way around, while the boxed-expression version supports both directions. Should we keep this distinction, or do we need to refactor?
> > 
> > 
> > Is there anything keeping us from handling both cases in a shared component?
> 
> We can cover that case as well. I previously discussed this with @barmac, who mentioned that only a one-way sync is required as of now.

The thing about this is that in the future, I expect that we replace dmn-js-literal-expression with dmn-js-boxed-expression entirely where you cannot change the variable name by design. Still, I think that for the external changes it makes sense to sync it two-way.

abdul99ahad commented:
> Note that the variable name change is not reflected in the literal expression editor:
> 
>  Screen.Recording.2024-09-13.at.10.40.49.mov

Yes, the variable name is changing but the state is not updating. Hence, when you're on the same screen (lieral expression editor), the change doesn't reflect although its value has been updated. 

For variable name changing directly, the element name won't change since it's currently not supported.

barmac commented:
> Yes, the variable name is changing but the state is not updating. Hence, when you're on the same screen (lieral expression editor), the change doesn't reflect although its value has been updated.

IMO the update should happen as it does in boxed expression. Let the field display the current value.

abdul99ahad commented:
> > Yes, the variable name is changing but the state is not updating. Hence, when you're on the same screen (lieral expression editor), the change doesn't reflect although its value has been updated.
> 
> IMO the update should happen as it does in boxed expression. Let the field display the current value.

@barmac need your support in updating the state externally as I'm stuck and can only update the state within the component. We might need to use Wrapper with ChangeSupport just like in boxed-expression.

barmac commented on packages/dmn-js-literal-expression/src/features/literal-expression-properties/components/LiteralExpressionPropertiesEditorComponent.js:
```diff
@@ -32,6 +32,21 @@ export default class LiteralExpressionPropertiesComponent extends Component {
 
+  componentWillMount() {
+    this._eventBus.on('elements.changed', this.onChange);
+    console.log('componentDidUpdate',);
```
```suggestion
```

barmac commented on packages/dmn-js-literal-expression/src/features/literal-expression-properties/components/LiteralExpressionPropertiesEditorComponent.js:
```diff
@@ -32,6 +32,21 @@ export default class LiteralExpressionPropertiesComponent extends Component {
+
+  onChange = () => {
+    this.setState({
+      name: this._viewer.getDecision().variable.name
```
How will this work in case no variable is added to a decision? I am speaking of cases like:

```xml
  <decision id="Decision_1kfrmu8" name="Decision 1">
    <literalExpression id="LiteralExpression_0lr3hs5">
      <text>a</text>
    </literalExpression>
  </decision>
```

While it cannot be created in our tool, it's a valid DMN which would have failed due to missing variable.

abdul99ahad replied:
In this case, should the text be treated as a variable, or should we ignore the variable and avoid updating the state?

barmac replied:
Ideally, I'd expect a variable to be created when we change the name or type. I see that it's probably scope creep since we don't handle cases like [above](https://github.com/bpmn-io/dmn-js/pull/893#discussion_r1771043404). So for now, I'd only expect that we don't add another place where the component may crash. Code-wise, this means that I'd not update the state if a variable is missing completely.

barmac commented:
The CI failure seems to be unrelated. I'll give it a second try.

barmac commented on packages/dmn-js-literal-expression/src/features/literal-expression-properties/components/LiteralExpressionPropertiesEditorComponent.js:
```diff
@@ -32,6 +32,23 @@ export default class LiteralExpressionPropertiesComponent extends Component {
+
+  onChange = () => {
+    const decision = this._viewer.getDecision();
+    if (decision.variable) {
```
🚀

barmac commented on packages/dmn-js-drd/src/features/modeling/cmd/UpdatePropertiesHandler.js:
```diff
@@ -37,8 +39,8 @@ UpdatePropertiesHandler.$inject = [ 'elementRegistry', 'moddle' ];
 
-  var element = context.element,
-      changed = [ element ];
+  const { element, properties } = context,
```
[optional] I'd consider to either have all declarations in ES6 (const, let), or leave this file with `var`s only. Right now it looks weird.

barmac commented:
One optional improvement suggestion, but other than that it looks good.

nikku commented on packages/dmn-js-shared/src/features/modeling/behavior/NameChangeBehavior.js:
```diff
@@ -0,0 +1,87 @@
+  }
+
+  isElementVariable(element) {
+    const variable = element.get('variable');
```
This breaks when using this feature along with the [dmn-js-properties-panel](https://github.com/bpmn-io/dmn-js-properties-panel).

nikku replied:
Cf. https://github.com/bpmn-io/dmn-js-properties-panel/issues/103

barmac replied:
Ouch. I can reproduce it on the Modeler nightly too.
